### PR TITLE
Add functions `List.compare_length` and `List.compare_length_with`

### PR DESCRIPTION
**Pull request by lefessan**
A standard inefficiency pattern is the test `List.length list > N` where N is often 0, 1 or 2, while `List.length` will still iter on the whole list. We propose to introduce two new functions:
- `List.compare_length l1 l2` compares the lengths of the two lists
- `List.length_compare l n` compares the length of the list to an integer
  Both functions stops after the minimal number of iterations needed on the lists.

### CODE REVIEW

**Comment by lpw25** [comment]
I like both those functions, but I think the names are too similar. `compare_length` seems to fit quite well with what it does, but `length_compare` is not a particularly obvious name so perhaps rename that one.

**Comment by hcarty** [comment]
Maybe `check_length`?

**Comment by lefessan** [comment]
Could also be `compare_lengths` and `compare_length`, depending on the number of lists passed as arguments.

**Comment by lijunsong** [comment]
This seems trivial, but `;;` at the end of the two functions feel so odd in this file.

**Comment by bobzhang** [comment]
@lefessan I wrote those two functions several times, btw, can `compare_length` do a little more work, it returns an int instead, if < 0 then is false, otherwise return the length

**Comment by braibant** [comment]
> it returns an int instead, if < 0 then is false, otherwise return the length

I really think we should avoid this kind of encodings. If this behavior is important, let's use a (polymorphic) variant type.

Re. the naming of `length_compare`, what about `compare_length_with_int`?

**Comment by jberdine** [comment]
Name suggestion from the peanut gallery: `compare_lengths` and `compare_to_length`?

Also, if the difference in length is returned instead of only -1, 0, 1, wouldn't the inefficiency of traversing the whole list remain, defeating much of the point of these?

**Comment by lefessan** [comment]
I followed @jberdine naming suggestion. Encoding the length of the shortest list in the result would be possible, but not very friendly (i.e. the result would be 1+length to avoid the zero). 

Maybe an alternative to `List.compare_to_length` would be to provide a `List.length_smaller_than list n` that returns the length of the list if smaller than `n`, else returns `n`.

**Comment by gasche** [comment]
`shorter_than l1 l2` and `length_less_than l1 n` would also be explicit names.

**Comment by yallop** [comment]
I like @gasche's [suggestions](https://github.com/ocaml/ocaml/pull/760#issuecomment-244310663).  Both `shorter_than` and `length_less_than` encourage a less imperative way of thinking, and read nicely in code:

``` ocaml
match e with
  A x, B y when shorter_than x y -> ...
```

**Comment by lefessan** [comment]
@gasche The functions do not return booleans, but integers, like the `compare` function.

**Comment by gasche** [comment]
Right. Then the best I can think of right now is `compare_length` (to be read in the sense of Haskell's `compare `on` length` and `compare_length_with` -- which I now notice is very similar to Thomas `compare_length_with_int`.

Another function that could replace most use-cases of `compare_length_with` is `valid_index` (or `has_nth`).

**Comment by alainfrisch** [comment]
Thought experiment: what would it take to let users write `if List.length l < 100 then ...` and have it compiled the efficient way?  Could flambda (be extended to) optimize the code automatically based on the definition of `List.length`?  A more pedestrian way would be to turn `List.length` into an external primitive and add ad hoc rules in the compiler to detect forms that can be optimized.  Many more optimizations could be implemented on functions from the stdlib if the compiler knew about them (e.g. avoiding intermediate copies in `s1 ^ s2 ^ s3`).

**Comment by Drup** [comment]
@alainfrisch So basically GHC's fusion rules ? A lot of people are going to feel strongly about that. :)

**Comment by alainfrisch** [comment]
I'm not familiar with GHC fusion rules, but it seems they can be specified in "normal" user code.  I was referring to something more ad hoc (exposing some well-chosen functions from the stdlib as `external %` primitives to let the compiler know about them).  This is much less powerful, but much simpler and perhaps a bit less controversial.

**Comment by lefessan** [comment]
The problem with "fusion rules" is how to be sure they are going to apply to your particular case ? For example, `if List.length x < 100 then ...` is easy to detect, but what about `let l = List.length x in if l < 100 then ...` where `l` is never reused ? At the end, if you need to know what is detected and what is not, it might be easier to just use directly the optimized functions.

**Comment by gasche** [comment]
I think that having optimizations that reduce constant factors (for example a more efficient test for `str = ""`) is fair game, but I would be very careful with optimization that introduce algorithmic speedups. In fact it's not so much about constant-time vs. more, it's about preserving or not my model of what code is running.

With list operations, there is a crisp type definition that _determines_ how most operations are implemented (but not their stack space usage), and anyone, even beginners, can quickly guess that `List.length` cannot possibly not traverse the whole list, same for `fold`, `map` etc. If we start optimizing arbitrary operations, how are we going to let people reason about efficiency without having to check, for any possible optimizable pattern, whether it is optimized or not?

To summarize, I think optimizing `(length li) <comparison> <constant>` is not a good idea.

**Comment by lefessan** [comment]
Actually, since compiler plugins will be available in 4.04, why not write a plugin to do these optimizations, and let developers choose to use or not use it for a particular module.

**Comment by xavierleroy** [comment]
> Thought experiment: what would it take to let users write if List.length l < 100 then ... and have it compiled the efficient way?

Don't overestimate the cleverness of the compiler.  I can see a compiler that performs compile-time symbolic evaluation turning `List.length l < 100` into

```
  match l with [] -> true | _:: l ->
  match l with [] -> true | _:: l ->
  (* etc, etc *)
  match l with [] -> true | _ :: _ -> false
```

Re-rolling these 100 matches into a recursive function is another matter entirely, and more of the domain of a program synthesis tool than of an optimizing compiler.

**Comment by lefessan** [comment]
Following @gasche comment, I changed `compare_to_length` into `compare_length_with`, but I still prefer `compare_lengths` to `compare_length`. I don't like `valid_index`, as a name for a function over lists, as it seems to encourage using a list instead of an array for indexed operations.

**Comment by lefessan** [comment]
It seems that we have reached a consensus. I add a commit to update CHANGES, and I merge.

**Comment by yallop** [comment]
`compare_lengths` is too specific.  It would be better to have a function

``` ocaml
List.compare : ('a -> 'a -> int) -> 'a list -> 'a list -> int
```

(generalizing the type to support different element types, if you prefer)

and then define

``` ocaml
compare_lengths = List.compare (fun _ _ -> 0)
```

**Comment by lefessan** [comment]
I see your point, but I am pretty sure it would make the performance of `compare_lengths` much worse (except maybe with flambda). I would suggest to make an independant PR for it, that would not modify `compare_lengths`.

File: testsuite/tests/translprim/comparison_table.ml.reference [hunk 1]
```diff
@@ -232,7 +232,7 @@
                 (apply f (field 0 param) (field 1 param)))
             map =
               (function f l
-                (apply (field 12 (global List!)) (apply uncurry f)
+                (apply (field 14 (global List!)) (apply uncurry f)
```
Comment (avsm) [inline testsuite/tests/translprim/comparison_table.ml.reference:235]: was this intended to be committed with the rest of the patch? It appears unrelated.

Comment (gasche) [reply testsuite/tests/translprim/comparison_table.ml.reference:235]: I think this test exposes the offset of some functions (`List.map` here, I suppose) in the `List` module, so it may change when new definitions are added before the functions used.

Comment (lefessan) [reply testsuite/tests/translprim/comparison_table.ml.reference:235]: Yes, it is a fix for the testsuite needed by the patch, because it changes the offset of `List.map` in the signature.
